From 596c1f3e070c93403e77b620d28d16c55cd5c0fa Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Fri, 9 Aug 2024 01:34:42 +0200
Subject: [PATCH] Add Version{LowBits,ChainId,HasAuxPow,IsLegacy} helpers (#41)

These functions are helpful when handling aux PoW.

(cherry picked from commit be4758c645ae0058cd3111ca95695a19ba4e0d23)

diff --git a/src/primitives/auxpow.h b/src/primitives/auxpow.h
index d5aa57d048a9fba147e6f1f3a48e96085a012349..0fb6372617bf75e00f91f970254767bc732dd8b0 100644
--- a/src/primitives/auxpow.h
+++ b/src/primitives/auxpow.h
@@ -36,4 +36,37 @@ int32_t MakeVersionWithChainId(uint32_t nChainId, uint32_t nLowVersionBits);
  */
 int32_t VersionWithAuxPow(uint32_t nVersion, bool hasAuxPow);
 
+/**
+ * Extract the low version bits, which are interpreted as the "traditional"
+ * Bitcoin version. The upper bits are used to signal presence of AuxPow and
+ * to set the chain ID.
+ */
+inline uint32_t VersionLowBits(int32_t nVersion) {
+    return nVersion & (VERSION_AUXPOW_BIT - 1);
+}
+
+/**
+ * Extract the chain ID from the nVersion.
+ */
+inline uint32_t VersionChainId(int32_t nVersion) {
+    return uint32_t(nVersion) >> VERSION_CHAIN_ID_BIT_POS;
+}
+
+/**
+ * Check if the auxpow flag is set in nVersion.
+ */
+inline bool VersionHasAuxPow(int32_t nVersion) {
+    return nVersion & VERSION_AUXPOW_BIT;
+}
+
+/**
+ * Check whether this is a "legacy" block without chain ID.
+ */
+inline bool VersionIsLegacy(int32_t nVersion) {
+    return nVersion == 1
+           // Dogecoin: We have a random v2 block with no AuxPoW, treat as
+           // legacy
+           || nVersion == 2;
+}
+
 #endif // BITCOIN_PRIMITIVES_AUXPOW_H
diff --git a/src/test/dogecoin_block_version_tests.cpp b/src/test/dogecoin_block_version_tests.cpp
index 418423398be0ba44f333880d70056ed9e57139c9..325a3134fa3116f3cb5b294dc337c35349313e54 100644
--- a/src/test/dogecoin_block_version_tests.cpp
+++ b/src/test/dogecoin_block_version_tests.cpp
@@ -59,4 +59,49 @@ BOOST_AUTO_TEST_CASE(VersionWithAuxPow_test) {
     BOOST_CHECK_EQUAL(VersionWithAuxPow(0xffff01ab, true), 0xffff01ab);
 }
 
+BOOST_AUTO_TEST_CASE(VersionLowBits_test) {
+    BOOST_CHECK_EQUAL(VersionLowBits(0), 0);
+    BOOST_CHECK_EQUAL(VersionLowBits(1), 1);
+    BOOST_CHECK_EQUAL(VersionLowBits(0xab), 0xab);
+    BOOST_CHECK_EQUAL(VersionLowBits(0x100), 0);
+    BOOST_CHECK_EQUAL(VersionLowBits(0x6200ab), 0xab);
+    BOOST_CHECK_EQUAL(VersionLowBits(0x6201ab), 0xab);
+    BOOST_CHECK_EQUAL(VersionLowBits(0xffff0100), 0x00);
+    BOOST_CHECK_EQUAL(VersionLowBits(0xffff01ab), 0xab);
+}
+
+BOOST_AUTO_TEST_CASE(VersionChainId_test) {
+    BOOST_CHECK_EQUAL(VersionChainId(0), 0);
+    BOOST_CHECK_EQUAL(VersionChainId(1), 0);
+    BOOST_CHECK_EQUAL(VersionChainId(0xab), 0);
+    BOOST_CHECK_EQUAL(VersionChainId(0x100), 0);
+    BOOST_CHECK_EQUAL(VersionChainId(0x6200ab), AUXPOW_CHAIN_ID);
+    BOOST_CHECK_EQUAL(VersionChainId(0x6201ab), AUXPOW_CHAIN_ID);
+    BOOST_CHECK_EQUAL(VersionChainId(0xffff0100), 0xffff);
+    BOOST_CHECK_EQUAL(VersionChainId(0xffff01ab), 0xffff);
+}
+
+BOOST_AUTO_TEST_CASE(VersionHasAuxPow_test) {
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0), false);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(1), false);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0xab), false);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0x100), true);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0x6200ab), false);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0x6201ab), true);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0xffff0100), true);
+    BOOST_CHECK_EQUAL(VersionHasAuxPow(0xffff01ab), true);
+}
+
+BOOST_AUTO_TEST_CASE(VersionIsLegacy_test) {
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(1), true);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(2), true);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(3), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0x100), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0x6200ab), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0x6201ab), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0xffff0100), false);
+    BOOST_CHECK_EQUAL(VersionIsLegacy(0xffff01ab), false);
+}
+
 BOOST_AUTO_TEST_SUITE_END()
